src/bgw/scheduler.c

```diff
@@ -0,0 +1,402 @@
+	CommitTransactionCommand();
+
+	elog(DEBUG1, "scheduling job %d \"%s\"", sjob->job->fd.id, NameStr(sjob->job->fd.job_name));
+
+
+	sjob->handle = ts_bgw_job_start_worker(sjob->job);
+	status = WaitForBackgroundWorkerStartup(sjob->handle, &pid);
+	switch (status)
+	{
+		case BGWH_POSTMASTER_DIED:
+			bgw_scheduler_on_postmaster_death();
+			break;
+		case BGWH_STARTED:
+			/* all good */
+			break;
+		case BGWH_STOPPED:
+			sjob->handle = NULL;
+			StartTransactionCommand();
+			ts_bgw_job_stat_mark_end(sjob->job, false);
+			scheduled_bgw_job_mark_stopped(sjob);
+			CommitTransactionCommand();
+			break;
+		case BGWH_NOT_YET_STARTED:
+			/* should not be possible */
+			elog(ERROR, "unexpected bgworker state %d", status);
+			break;
+	}
+}
+
+static ScheduledBgwJob *
+scheduled_bgw_job_create(BgwJob *job, MemoryContext mctx)
+{
+	BgwJobStat *job_stat = ts_bgw_job_stat_find(job->fd.id);
+	ScheduledBgwJob *sjob = MemoryContextAlloc(mctx, sizeof(ScheduledBgwJob));
+
+	*sjob = (ScheduledBgwJob)
+	{
+		.job = job,
+			.enabled = ts_bgw_job_stat_should_execute(job_stat, job),
+			.next_start = ts_bgw_job_stat_next_start(job_stat, job),
+	};
+
+	return sjob;
+}
+
+static bool
+scheduled_bgw_job_is_running(ScheduledBgwJob *sjob)
+{
+	return sjob->handle != NULL;
+}
+
+static void
+scheduled_bgw_job_mark_stopped(ScheduledBgwJob *sjob)
+{
+	BgwJobStat *job_stat = ts_bgw_job_stat_find(sjob->job->fd.id);
+
+	if (!ts_bgw_job_stat_end_was_marked(job_stat))
+	{
+		/*
+		 * This happens not on a crash but rather upon a signal (cancel or
+		 * terminate) of the job
+		 */
+		ts_bgw_job_stat_mark_end(sjob->job, false);
+		/* refresh */
+		job_stat = ts_bgw_job_stat_find(sjob->job->fd.id);
+	}
+
+	sjob->handle = NULL;
+	sjob->next_start = ts_bgw_job_stat_next_start(job_stat, sjob->job);
+	sjob->enabled = ts_bgw_job_stat_should_execute(job_stat, sjob->job);
+}
+
+static
+List *
+scheduled_bgw_job_list_create(MemoryContext mctx)
+{
+	List	   *res = NIL;
+	List	   *jobs = ts_bgw_job_get_all(mctx);
+	ListCell   *lc;
+	MemoryContext old;
+
+	foreach(lc, jobs)
+	{
+		BgwJob	   *job = lfirst(lc);
+		ScheduledBgwJob *sjob = scheduled_bgw_job_create(job, mctx);
+
+		old = MemoryContextSwitchTo(mctx);
+		res = lappend(res, sjob);
+		MemoryContextSwitchTo(old);
+	}
+	return res;
+}
+
+static
+bool
+scheduled_bgw_job_list_start_workers()
+{
+	ListCell   *lc;
+	bool		running = false;
+
+	foreach(lc, scheduled_jobs)
+	{
+		ScheduledBgwJob *sjob = lfirst(lc);
+
+		if (scheduled_bgw_job_is_running(sjob))
+		{
+			running = true;
+			continue;
+		}
+		if (!sjob->enabled || get_timeout(sjob->next_start) > 0)
+			continue;
+		scheduled_bgw_job_start_worker(sjob);
+		if (sjob->handle != NULL)
+			running = true;
+	}
+	return running;
+}
+
+static TimestampTz
+scheduled_bgw_job_list_next_start(TimestampTz next_start)
+{
+	ListCell   *lc;
+
+	foreach(lc, scheduled_jobs)
+	{
+		ScheduledBgwJob *sjob = lfirst(lc);
+
+		if (!sjob->enabled || scheduled_bgw_job_is_running(sjob))
```

Comment:
`scheduled_bgw_job_is_running` looks like it can cause deadlocks in the case where a background worker starts and immediately finishes before the scheduler gets a chance to check if the job is still alive. We should mention that in such a case the `WaitLatch` will be set by the default `SIGUSR1` handler.

Reply:
Not sure I understand scheduled_bgw_job_is_running checks a local variable. What do you mean?